File: ws/first_ws/src/topic_pkg/scripts/custom_msg_obj_subscriber.py
#! /usr/bin/env python
# coding:utf8

import rospy
from demo_msgs.msg import Student, Team


def topic_callback(msg):
    if not isinstance(msg, Team):
        return
    print "==="
    print "team::name %s" % msg.name
    print "team::leader::name %s" % msg.leader.name
    print "team::leader::age %d" % msg.leader.age


if __name__ == '__main__':
    nodeName = "subscribe_custom_msg_py_node"
    rospy.init_node(nodeName)

    topicName = "/custom/msg/obj/topic"
    rospy.Subscriber(topicName, Team, topic_callback)

    rospy.spin()

File: ws/first_ws/src/topic_pkg/src/custom_msg_obj_subscriber.cpp
//
// Created by zcb on 2021/4/2.
//

#include <iostream>
#include <ros/ros.h>
#include <demo_msgs/Team.h>

using namespace std;

void topicCallBack(const demo_msgs::Team::ConstPtr& msg){
    cout << "===" << endl;
    cout << "team::name "<<msg->name<<endl;
    cout << "team::leader::name " << msg->leader.name << endl;
    cout << "team::leader::age " << msg->leader.age << endl;
}


int main(int argc,char** argv){
    string nodeName = "subscribe_custom_msg_obj_cpp_node";
    ros::init(argc,argv,nodeName);
    ros::NodeHandle node;

    string topicName = "/custom/msg/obj/topic";
    ros::Subscriber sub = node.subscribe(topicName,1000,topicCallBack);

    ros::spin();

    return 0;
}

